okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java

```diff
@@ -0,0 +1,294 @@
+/*
+ * Copyright (C) 2015 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.okhttp;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.URI;
+import java.net.URL;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A <a href="https://url.spec.whatwg.org/">URL</a> with an {@code http} or {@code https} scheme.
```

Comment:
Agreed.

The thing I like about this approach is that it's 1:1 with the URLs supported by our request/response model. It'd be weird if a page could redirect to `wss://localhost:80` and we attempted to traverse that link.

I reserve the right to change my mind on this later.